tests/phpunit/tests/option/option.php

```diff
@@ -312,4 +312,151 @@ public function data_option_autoloading() {
 			array( 'autoload_false', false, 'no' ),
 		);
 	}
+
+	/**
+	 * @ticket 22192
+	 *
+	 * @covers ::add_option
+	 */
+	public function test_add_option_with_value_of_false_should_store_false_in_the_cache() {
+		add_option( 'foo', false );
+		$a = wp_cache_get( 'alloptions', 'options' );
+		$this->assertSame( false, $a['foo'] );
+	}
+
+	/**
+	 * @ticket 22192
+	 *
+	 * @covers ::add_option
+	 */
+	public function test_add_option_with_value_of_false_should_store_empty_string_in_the_database() {
+		add_option( 'foo', false );
+
+		// Delete cache to ensure we pull from the database.
+		wp_cache_delete( 'alloptions', 'options' );
+
+		$this->assertSame( '', get_option( 'foo' ) );
+	}
+
+	/**
+	 * @ticket 22192
+	 *
+	 * @covers ::add_option
+	 * @covers ::update_option
+	 *
+	 * @dataProvider data_update_option_type_juggling
+	 *
+	 * @param mixed $old_value One of the values to compare.
+	 * @param mixed $new_value The other value to compare.
+	 */
+	public function test_update_option_should_hit_cache_when_loosely_equal_to_existing_value_and_cached_values_are_faithful_to_original_type( $old_value, $new_value ) {
+		add_option( 'foo', $old_value );
+		$num_queries = get_num_queries();
+
+		$updated = update_option( 'foo', $new_value );
+
+		$this->assertFalse( $updated, 'update_option should not return true when values are loosely equal.' );
+		$this->assertSame( $num_queries, get_num_queries(), 'The number of database queries should not change.' );
+	}
+
+	/**
+	 * @ticket 22192
+	 *
+	 * @covers ::add_option
+	 * @covers ::update_option
+	 *
+	 * @dataProvider data_update_option_type_juggling
+	 *
+	 * @param mixed $old_value One of the values to compare.
+	 * @param mixed $new_value The other value to compare.
+	 */
+	public function test_update_option_should_hit_cache_when_loosely_equal_to_existing_value_and_cached_values_are_pulled_from_the_database( $old_value, $new_value ) {
+		add_option( 'foo', $old_value );
+		wp_cache_delete( 'alloptions', 'options' );
+		wp_load_alloptions();
+
+		$num_queries = get_num_queries();
+
+		$updated = update_option( 'foo', $new_value );
+
+		$this->assertFalse( $updated, 'update_option should not return true when values are loosely equal.' );
+		$this->assertSame( $num_queries, get_num_queries(), 'The number of database queries should not change.' );
+	}
+
+	/**
+	 * Data provider.
+	 *
+	 * @return array
+	 */
+	public function data_update_option_type_juggling() {
+		return array(
+			// Truthy.
+			array( '1', '1' ),
+			array( '1', intval( 1 ) ),
```

Comment:
We can remove the `intval()` and `floatval()` functions here and just replace with `1` and `1.0` respectively.

:1234: Applies to other datasets.